Multibody actuators

amcastro-tri opened the pull request:
Implements `JointActuators` which essentially is just bookkeeping sugar to keep track of what joints are actuated. MBP uses them to setup an input port for actuation and apply actuation on each joint.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8147)
<!-- Reviewable:end -->

amcastro-tri commented:
+@edrumwri for feature review please.

---

Review status: 0 of 14 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-:-L6-QDsCQnFu81-3lIfx:b-a98c4g)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Reviewed 14 of 21 files at r2.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[multibody/benchmarks/acrobot/make_acrobot_plant.cc, line 68 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62RZZ7IwF2WggeYWZ4:-L62RZZ7IwF2WggeYWZ5:bvcgamu) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/benchmarks/acrobot/make_acrobot_plant.cc#L68)):*
> ```cpp
>       Vector3d::UnitY()); /* acrobot oscillates in the x-z plane. */
> 
>   const RevoluteJoint<double>& elbow = plant->AddJoint<RevoluteJoint>(
> ```

It would be nice to have some sugar here so that you could create the joint and it's actuator at the same time.

---

*[multibody/benchmarks/acrobot/make_acrobot_plant.h, line 119 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62RIxIxQcXEVGg4T6p:-L62RIxIxQcXEVGg4T6q:b67d4dt) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/benchmarks/acrobot/make_acrobot_plant.h#L119)):*
> ```objc
> 
>   // Modeling elements' names.
>   std::string link1_name_{"Link1"};
> ```

BTW Make these `const`?

---

*[multibody/benchmarks/kuka_iiwa_robot/make_kuka_iiwa_model.cc, line 111 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62RzDmIQEvA6REw3VW:-L62RzDmIQEvA6REw3VX:bfuqklc) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/benchmarks/kuka_iiwa_robot/make_kuka_iiwa_model.cc#L111)):*
> ```cpp
>   // angles and a position vector. Alternately, frame An is regarded as
>   // coincident with linkA.
>   const Joint<T>* joint{nullptr};
> ```

Having to reuse `joint` multiple times begs for a different construction approach.

---

*[multibody/multibody_tree/joint_actuator.h, line 22 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62T11-23jEeR9QLYii:-L62T11-23jEeR9QLYij:bkds1x2) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/joint_actuator.h#L22)):*
> ```objc
> /// represent an actuator acting on a given Joint. When added to a MultibodyTree
> /// model, a %JointActuator gets assigned a JointActuatorIndex, which can be
> /// retrieved with JointActuator::get_index(). The %JointActuator class
> ```

"JointActuator::get__index()" -> "`JointActuator::get__index()`"

---

*[multibody/multibody_tree/joint_actuator.h, line 47 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62VIuWmy8YemOkBkIY:-L62VIuWmy8YemOkBkIZ:b-ss5jdx) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/joint_actuator.h#L47)):*
> ```objc
> 
>   /// Returns the name of the actuator.
>   const std::string& get_name() const { return name_; }
> ```

BTW Recommend `name()` for consistency with `joint()` below.

---

*[multibody/multibody_tree/joint_actuator.h, line 87 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62V0GfXQ4GhdFFQISa:-L62V0GfXQ4GhdFFQISb:bpbbqe5) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/joint_actuator.h#L87)):*
> ```objc
>   // from the parent MultibodyTree.
>   void DoSetTopology(const MultibodyTreeTopology&) final {
>     // So far this initial implementation has no topology counterpart.
> ```

@amcastro-tri, I'm not sure what you're trying to tell me with this comment.

---

*[multibody/multibody_tree/joint_actuator.h, line 91 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62VAMRMgJp1hQRx67i:-L62VAMRMgJp1hQRx67j:b-1rwov2) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/joint_actuator.h#L91)):*
> ```objc
> 
>   // The actuator's unique name in the MultibodyTree model
>   std::string name_;
> ```

BTW `const`?

---

*[multibody/multibody_tree/joint_actuator.h, line 94 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62VE5507Ww8yEXLEsN:-L62VE5507Ww8yEXLEsO:b-1rwov2) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/joint_actuator.h#L94)):*
> ```objc
> 
>   // The index of the joint on which this actuator acts.
>   JointIndex joint_index_;
> ```

BTW `const`?

---

*[multibody/multibody_tree/multibody_tree.h, line 567 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62WfVRyKq5wKm8ccKb:-L62WfVSvhW0ro_StXmI:b-67sw8a) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/multibody_tree.h#L567)):*
> ```objc
>   /// @param[in] name
>   ///   A string that uniquely identifies the new actuator to be added to `this`
>   ///   model. A std::runtime_error is thrown if an actuator with the same name
> ```

Exception types don't match.

---

*[multibody/multibody_tree/multibody_tree.h, line 575 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62WzskXeM_vKNfcHIA:-L62WzslTNm-oj0KFHes:b-lwybj7) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/multibody_tree.h#L575)):*
> ```objc
>   const JointActuator<T>& AddJointActuator(
>       const std::string& name, const Joint<T>& joint) {
>     if (HasJointActuatorNamed(name)) {
> ```

It appears that you're doing an O(n lg n) operation here- actual complexity is O(lg 1 + lg 2 + ... lg n). You either should document that this operation is expensive or- better (imo)- verify that all actuator names are unique at finalization of the tree.

---

*[multibody/multibody_tree/multibody_tree.h, line 691 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62VY3Mai735gVQhrWx:-L62VY3N_1-lwygA_DqU:b-owxsys) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/multibody_tree.h#L691)):*
> ```objc
>   /// Returns a constant reference to the joint actuator with unique index
>   /// `actuator_index`.
>   /// @throws std::runtime_error when `actuator_index` does not correspond to a
> ```

BTW I'd recommend changing this remove the specific exception type, since DRAKE_THROW_UNLESS() could change the type of exception it throws, causing this documentation to become out-of-date.

---

*[multibody/multibody_tree/multibody_tree.h, line 812 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62Vy-RVBNbQVWVYM8L:-L62Vy-RVBNbQVWVYM8M:b-gqznvw) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/multibody_tree.h#L812)):*
> ```objc
>   const JointActuator<T>& GetJointActuatorByName(
>       const std::string& name) const {
>     auto it = actuator_name_to_index_.find(name);
> ```

BTW `auto` -> `const auto`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[multibody/benchmarks/acrobot/make_acrobot_plant.cc, line 68 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62RZZ7IwF2WggeYWZ4:-L6J40BAe7uRVeC6A4Wh:bstdryl) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/benchmarks/acrobot/make_acrobot_plant.cc#L68)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

It would be nice to have some sugar here so that you could create the joint and it's actuator at the same time.
</blockquote></details>

That's an interesting idea. How would you like or how would you think this line would read with an API like that?

---

*[multibody/benchmarks/kuka_iiwa_robot/make_kuka_iiwa_model.cc, line 111 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62RzDmIQEvA6REw3VW:-L6J4Ek-5spzXzimHBU8:bip9mce) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/benchmarks/kuka_iiwa_robot/make_kuka_iiwa_model.cc#L111)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Having to reuse `joint` multiple times begs for a different construction approach.
</blockquote></details>

I could have a const reference at each of these lines if you think that reads better?

---

*[multibody/multibody_tree/joint_actuator.h, line 22 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62T11-23jEeR9QLYii:-L6J5H0h8QkW66T23d7k:b-thrrfu) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/joint_actuator.h#L22)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

"JointActuator::get__index()" -> "`JointActuator::get__index()`"
</blockquote></details>

sorry, are you making reference to a double underscore? it is a single underscore actually if that's what you mean.

---

*[multibody/multibody_tree/joint_actuator.h, line 47 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62VIuWmy8YemOkBkIY:-L6J5SFuTVPZpfDqFVRu:by2h9or) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/joint_actuator.h#L47)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW Recommend `name()` for consistency with `joint()` below.
</blockquote></details>

BTW, yes, I need to do a cleanup on getters. Before we were preferring the `get_something()` until we realized `something()` does actually as well and requires less typing. I'll change to `name()` as you indicate and do a cleanup pass in a separate PR.

---

*[multibody/multibody_tree/multibody_tree.h, line 575 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62WzskXeM_vKNfcHIA:-L6J4iS1Gyar7ZZ4qbRJ:b-7bckum) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/multibody_tree.h#L575)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

It appears that you're doing an O(n lg n) operation here- actual complexity is O(lg 1 + lg 2 + ... lg n). You either should document that this operation is expensive or- better (imo)- verify that all actuator names are unique at finalization of the tree.
</blockquote></details>

This is only to add actuators when you are creating the model, not really a critical computational bottleneck and that is why I don't care much about it. Not at least until it does become a real problem. I tend to prefer reporting an error as soon as it happens vs waiting to the finalize stage to do so.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[multibody/multibody_tree/multibody_tree.h, line 575 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62WzskXeM_vKNfcHIA:-L6JsLMrlXGqEAci4RPR:b-ayhok4) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/multibody_tree.h#L575)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This is only to add actuators when you are creating the model, not really a critical computational bottleneck and that is why I don't care much about it. Not at least until it does become a real problem. I tend to prefer reporting an error as soon as it happens vs waiting to the finalize stage to do so.
</blockquote></details>

That's a reasonable design decision, but it's also counterintuitive- I would expect that when I add a joint actuator that it would be a constant time operation. So I'm simply saying that you should document it publicly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 11 of 14 files reviewed at latest revision, 8 unresolved discussions.

---

*[multibody/benchmarks/acrobot/make_acrobot_plant.cc, line 68 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62RZZ7IwF2WggeYWZ4:-L6RrSi8S_jbg0TaiARA:b-v3llmp) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/benchmarks/acrobot/make_acrobot_plant.cc#L68)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

That's an interesting idea. How would you like or how would you think this line would read with an API like that?
</blockquote></details>

Would that be additional "sugar" on top of the API proposed here? or a completely different API? We can always add sugar later. Right now we just need a mechanism to declare actuators. 

---

*[multibody/benchmarks/kuka_iiwa_robot/make_kuka_iiwa_model.cc, line 111 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62RzDmIQEvA6REw3VW:-L6Rrjv1y0tws5LTD6ea:b-81vo02) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/benchmarks/kuka_iiwa_robot/make_kuka_iiwa_model.cc#L111)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I could have a const reference at each of these lines if you think that reads better?
</blockquote></details>

Do you still think having a different const ref per line is a better approach? Pointer `joint` here is just playing the role of a local dummy variable.

---

*[multibody/multibody_tree/joint_actuator.h, line 87 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62V0GfXQ4GhdFFQISa:-L6RmQ-oHefRW6VD58sz:b-m7hzf1) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/joint_actuator.h#L87)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

@amcastro-tri, I'm not sure what you're trying to tell me with this comment.
</blockquote></details>

Removed to avoid confusion.

---

*[multibody/multibody_tree/joint_actuator.h, line 91 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62VAMRMgJp1hQRx67i:-L6RnE4D75AOPfE75ymh:b-fhv8hc) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/joint_actuator.h#L91)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW `const`?
</blockquote></details>

BTW see comment below. Even when in this case I could make it const, I don't see the value since users will never have mutable access to this member not even to this object.

---

*[multibody/multibody_tree/joint_actuator.h, line 94 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62VE5507Ww8yEXLEsN:-L6Rm_xDIY4wns5USIKP:bwqnzc7) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/joint_actuator.h#L94)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW `const`?
</blockquote></details>

BTW I need this one to be non-const so that MultibodyTree can assign it after its constructor is invoked. From a user's point of view, there is no mutable access to this member and even more, they only get back a const reference to the newly added actuator after calling `MBT::AddJointActuator()`, which is the only mechanism they have to create and add them. This is a constant them in the MBT API and it worked out really well.

---

*[multibody/multibody_tree/multibody_tree.h, line 567 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62WfVRyKq5wKm8ccKb:-L6Ro2K02mgss7u1qG8u:boj1qgt) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/multibody_tree.h#L567)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Exception types don't match.
</blockquote></details>

Good catch. Done.

---

*[multibody/multibody_tree/multibody_tree.h, line 575 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62WzskXeM_vKNfcHIA:-L6RpnmFrrhfWByQKXfe:bxki8r0) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/multibody_tree.h#L575)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

That's a reasonable design decision, but it's also counterintuitive- I would expect that when I add a joint actuator that it would be a constant time operation. So I'm simply saying that you should document it publicly.
</blockquote></details>

Notice @edrumwri that I am using `std::unordered_map` (constant average case) not a `std::map` (logarithmic complexity as you mention). Therefore the average complexity of the addition of n actuators will be `O(n)`. Yes, worst case scenario would be `O(n^2)` since the worst case complexity of an `std::unordered_map:find()` is linear with container size. If we believe the average complexity is the most common case, then I believe there's nothing to add, since clearly as you say,  a user would expect adding n joints costing something linear with n. 
However, do you think it is worth documenting the worst case complexity? Would it be a good idea to add worst case complexity doc for the `HasXXX()` methods? for the case where the key is a `std::string`, is there anything else we could say or promise? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Thanks for the review @edrumwri, back to you.

---

Review status: 11 of 14 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-:-L6RsHGRwdJjjkSmh53Q:b-q3hu72)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[multibody/benchmarks/acrobot/make_acrobot_plant.cc, line 68 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62RZZ7IwF2WggeYWZ4:-L6Ru3NivW7EQc_W9aoV:b5bhtnm) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/benchmarks/acrobot/make_acrobot_plant.cc#L68)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Would that be additional "sugar" on top of the API proposed here? or a completely different API? We can always add sugar later. Right now we just need a mechanism to declare actuators. 
</blockquote></details>

Right- the sugar can come later if you like. A TODO would suit me fine. Just keep in mind that your construction in make_kuka_iiwa_model.cc is currently suffering from this lack of sugar.

---

*[multibody/benchmarks/kuka_iiwa_robot/make_kuka_iiwa_model.cc, line 111 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62RzDmIQEvA6REw3VW:-L6RuJqNtprEcHJEY_A4:bv90g3q) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/benchmarks/kuka_iiwa_robot/make_kuka_iiwa_model.cc#L111)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Do you still think having a different const ref per line is a better approach? Pointer `joint` here is just playing the role of a local dummy variable.
</blockquote></details>

Right- my original comment implied that the sugar I was seeking would address this problem. If you're not doing that for this PR, feel free to ignore this comment. But as it stands, it's a little ugly.

---

*[multibody/multibody_tree/joint_actuator.h, line 22 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62T11-23jEeR9QLYii:-L6NaLIO5SLXrFZGo28B:b-nrhjj3) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/joint_actuator.h#L22)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

sorry, are you making reference to a double underscore? it is a single underscore actually if that's what you mean.
</blockquote></details>

Sorry, this was clearly a typo. I meant to say to change it to "::index()" for consistency.

---

*[multibody/multibody_tree/joint_actuator.h, line 47 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62VIuWmy8YemOkBkIY:-L6RucRUZ4QSSTEIf7-7:bub8141) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/joint_actuator.h#L47)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, yes, I need to do a cleanup on getters. Before we were preferring the `get_something()` until we realized `something()` does actually as well and requires less typing. I'll change to `name()` as you indicate and do a cleanup pass in a separate PR.
</blockquote></details>

Ok, then please mark it in some way as such (a TODO would suit me) so that it's not forgettable.

---

*[multibody/multibody_tree/joint_actuator.h, line 91 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62VAMRMgJp1hQRx67i:-L6RuwP5CTMAL7OF0Vg1:b-93anie) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/joint_actuator.h#L91)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW see comment below. Even when in this case I could make it const, I don't see the value since users will never have mutable access to this member not even to this object.
</blockquote></details>

No problem. There's usually little extrinsic "value" to making it const- it's a prophylatic measure.

---

*[multibody/multibody_tree/joint_actuator.h, line 94 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62VE5507Ww8yEXLEsN:-L6RuuIYKGqkrIVtMA7x:blyphgp) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/joint_actuator.h#L94)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW I need this one to be non-const so that MultibodyTree can assign it after its constructor is invoked. From a user's point of view, there is no mutable access to this member and even more, they only get back a const reference to the newly added actuator after calling `MBT::AddJointActuator()`, which is the only mechanism they have to create and add them. This is a constant them in the MBT API and it worked out really well.
</blockquote></details>

No problem.

---

*[multibody/multibody_tree/multibody_tree.h, line 575 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62WzskXeM_vKNfcHIA:-L6RvCVNCcPWwxk-RSbr:b-8voo1w) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/multibody_tree.h#L575)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Notice @edrumwri that I am using `std::unordered_map` (constant average case) not a `std::map` (logarithmic complexity as you mention). Therefore the average complexity of the addition of n actuators will be `O(n)`. Yes, worst case scenario would be `O(n^2)` since the worst case complexity of an `std::unordered_map:find()` is linear with container size. If we believe the average complexity is the most common case, then I believe there's nothing to add, since clearly as you say,  a user would expect adding n joints costing something linear with n. 
However, do you think it is worth documenting the worst case complexity? Would it be a good idea to add worst case complexity doc for the `HasXXX()` methods? for the case where the key is a `std::string`, is there anything else we could say or promise? 

</blockquote></details>

My mistake- I saw unordered_map and thought map. I think the expected time is fine.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[multibody/benchmarks/acrobot/make_acrobot_plant.cc, line 68 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62RZZ7IwF2WggeYWZ4:-L6SCKTM_AWZduxHmoy5:b-ojh0lw) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/benchmarks/acrobot/make_acrobot_plant.cc#L68)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Right- the sugar can come later if you like. A TODO would suit me fine. Just keep in mind that your construction in make_kuka_iiwa_model.cc is currently suffering from this lack of sugar.
</blockquote></details>

Done, added TODO in `MBT::AddJointActuator()`.

---

*[multibody/benchmarks/acrobot/make_acrobot_plant.h, line 119 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62RIxIxQcXEVGg4T6p:-L6SDjhIZvujKj3CDng6:bwdoxnm) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/benchmarks/acrobot/make_acrobot_plant.h#L119)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW Make these `const`?
</blockquote></details>

BTW, I reverted these to non-const since `const` doesn't play well with the default copy constructor provided through `DRAKE_DEFAULT_COPY_AND_MOVE_AND_ASSIGN`.

---

*[multibody/multibody_tree/joint_actuator.h, line 22 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62T11-23jEeR9QLYii:-L6SCba8UeLi2vlMA8Jz:b-h8jjpf) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/joint_actuator.h#L22)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Sorry, this was clearly a typo. I meant to say to change it to "::index()" for consistency.
</blockquote></details>

I just can't easily, `get_index()` gets inherited from `MultibodyTreeElement`. I was thinking to do a cleanup PR replacing all `get_foo()` methods by `foo()` methods now we know we like those better. But if I do it here you'll see a lot of refactoring in all places where it gets called. Allow me to do it in a follow PR (I can do that as soon as today, since clearly I put it off for too long.)

---

*[multibody/multibody_tree/joint_actuator.h, line 47 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62VIuWmy8YemOkBkIY:-L6SD6XsO27_T9voLH8k:bxgqumk) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/joint_actuator.h#L47)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Ok, then please mark it in some way as such (a TODO would suit me) so that it's not forgettable.
</blockquote></details>

I will submit a PR as soon as today. We'll see which one lands first.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
+@seancurtis-tri for platform review please.

---

Review status: 12 of 14 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-:-L6SEJO9lBKeBrj--1pO:b-tnzbcp)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-:-L6SVNcmrvmeLw4rwUrW:bnfp4nl)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
First pass done

---

Reviewed 11 of 21 files at r2, 1 of 3 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[multibody/multibody_tree/joint_actuator.h, line 47 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62VIuWmy8YemOkBkIY:-L6SXR_5XfaKiPAJb5nJ:b-p9to9m) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/joint_actuator.h#L47)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I will submit a PR as soon as today. We'll see which one lands first.
</blockquote></details>

Given this is a brand new file in this PR, there's no reason to introduce this inconsistency. You should change it to `name()` in the scope of this PR.

---

*[multibody/multibody_tree/joint_actuator.h, line 24 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6SWgdA0ViGebOk_ovO:-L6SWgdA0ViGebOk_ovP:b-uw5zff) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/joint_actuator.h#L24)):*
> ```objc
> /// retrieved with JointActuator::get_index(). The %JointActuator class
> /// allows mapping this JointActuatorIndex to the Joint on which it actuates,
> /// which can be retrieved with JointActuator::joint().
> ```

BTW This documentation is a bit confusing. The existence of a `JointActuatorIndex` is purely an internal implementation detail. There are no public methods to access it. So, it seems odd that it is featured so heavily in the doxygen. It seems to be more consistent with internal documentation.

---

*[multibody/multibody_tree/joint_actuator.h, line 35 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6SYga_2tL2VTILd_p8:-L6SYga_2tL2VTILd_p9:brtjrdg) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/joint_actuator.h#L35)):*
> ```objc
> /// They are already available to link against in the containing library.
> /// No other values for T are currently supported.
> template <typename T>
> ```

FYI The only reason this class is templated is so you can return a `Joint<T>`. Had you considered other options?  E.g.,

1. MultiBodyTree::GetActuatedJoint(const JointActuator& actuator);
2. actuator.GetJoint<T>();

---

*[multibody/multibody_tree/joint_actuator.h, line 43 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6SXEgAY-MtudiAhjsN:-L6SXEgAY-MtudiAhjsO:bx1i6ex) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/joint_actuator.h#L43)):*
> ```objc
>   /// Creates an actuator for `joint` with the given `name`.
>   /// The name must be unique within the given MultibodyTree model. This is
>   /// guaranteed by MultibodyTree::AddJointActuator().
> ```

BTW s/guaranteed/enforced

---

*[multibody/multibody_tree/multibody_tree.h, line 561 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6S_Se95r3WbMWooJGU:-L6S_Se95r3WbMWooJGV:b-53q6fe) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/multibody_tree.h#L561)):*
> ```objc
> 
>   /// Creates and adds a JointActuator model for an actuator acting on a given
>   /// joint.
> ```

BTW back ticks on `joint`.

---

*[multibody/multibody_tree/multibody_plant/BUILD.bazel, line 25 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6Sc4AVa6JJXdWHBKG-:-L6Sc4AVa6JJXdWHBKG0:b-nfc8vw) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/multibody_plant/BUILD.bazel#L25)):*
> ```Python
>     deps = [
>         "//drake/common:default_scalars",
>         "//drake/geometry:geometry_ids",
> ```

Given this is ostensibly about adding actuators, why are changes like these included?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 144 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6Sawb09z00job1SuHu:-L6Sawb09z00job1SuHv:b-qu7hf0) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L144)):*
> ```cpp
> const systems::InputPortDescriptor<T>&
> MultibodyPlant<T>::get_actuation_input_port() const {
>   return systems::System<T>::get_input_port(actuation_port_);
> ```

BTW Do you like the error message on this if no actuation was added?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 293 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6SZjrs5j6yNbKRyeRp:-L6SZjrs5j6yNbKRyeRq:b-2zx89k) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/multibody_plant/multibody_plant.h#L293)):*
> ```objc
> 
>   /// Creates and adds a JointActuator model for an actuator acting on a given
>   /// joint.
> ```

BTW backticks on `joint`.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 383 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6S_1_LNzsfH1HORRQG:-L6S_1_LNzsfH1HORRQH:b-a4wej6) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/multibody_plant/multibody_plant.h#L383)):*
> ```objc
> 
>   /// Returns a constant reference to the input port for external actuation.
>   /// This input port is a vector valued port, indexed by JointActuatorIndex.
> ```

BTW It isn't clear to me that `JointActuatorIndex` is a part of the public API. Where would the user get such a value to use as a meaningful index?

Also, how does this work with a joint with multiple dof? In the implementation of the forces, you note that only 1-dof forces are supported. So, that should probably be included elsewhere in the documentation (here, the constructor for `JointActuator`, etc. And it should likewise be enforced.)

---

*[multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc, line 124 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6ScVdrjT1PK-_nuCYA:-L6ScVdrjT1PK-_nuCYB:bm79648) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc#L124)):*
> ```cpp
> 
>     input_port_ = &context_->FixInputPort(
>         plant_->get_actuation_input_port().get_index(), Vector1<double>(0.0));
> ```

This should probably be preceded by an `ASSERT_GT(plant_->num_actutors(), 0);`.

---

*[multibody/multibody_tree/test/multibody_tree_test.cc, line 63 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6ScmK-jWfFniHLiL0G:-L6ScmK-jWfFniHLiL0H:b-yxdgaf) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/test/multibody_tree_test.cc#L63)):*
> ```cpp
>       "iiwa_actuator_7"};
> 
>   // Model Size. Counting the world body, there should be three bodies.
> ```

BTW copypasta "three bodies" should be "eight"

---

*[multibody/multibody_tree/test/multibody_tree_test.cc, line 73 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6ScsrNMq1l4r_WQncZ:-L6ScsrNMq1l4r_WQnc_:b-zhbqer) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/test/multibody_tree_test.cc#L73)):*
> ```cpp
>   EXPECT_EQ(model.get_num_states(), 14);
> 
>   // Query if elements exist in the model.
> ```

FYI s/elements/bodies

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-:-L6SdsT6226XFKn8JVNs:b1ccr98)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[multibody/multibody_tree/joint_actuator.h, line 47 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L62VIuWmy8YemOkBkIY:-L6StiPQUOoj6hRfETt-:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7c27c88f5927f7e0f4bc41c6eaac9a15c1eda929/multibody/multibody_tree/joint_actuator.h#L47)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Given this is a brand new file in this PR, there's no reason to introduce this inconsistency. You should change it to `name()` in the scope of this PR.
</blockquote></details>

Done.

---

*[multibody/multibody_tree/joint_actuator.h, line 24 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6SWgdA0ViGebOk_ovO:-L6SshoF_TsQB7rugP_q:b5rwptl) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/joint_actuator.h#L24)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW This documentation is a bit confusing. The existence of a `JointActuatorIndex` is purely an internal implementation detail. There are no public methods to access it. So, it seems odd that it is featured so heavily in the doxygen. It seems to be more consistent with internal documentation.
</blockquote></details>

BTW I think the docs I added now in MBP help a bit. Otherwise, it'll become clearer when I add the computation of the B matrix.

---

*[multibody/multibody_tree/joint_actuator.h, line 35 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6SYga_2tL2VTILd_p8:-L6St6yH1Je3OCN9RL0-:b-gqznxj) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/joint_actuator.h#L35)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI The only reason this class is templated is so you can return a `Joint<T>`. Had you considered other options?  E.g.,

1. MultiBodyTree::GetActuatedJoint(const JointActuator& actuator);
2. actuator.GetJoint<T>();
</blockquote></details>

I thought about it, yes. I settled here just to provide a homogeneous API, all MBT modeling elements are templated on `<T>`. Also, this would allow us to add `Context` methods in here if we decide. Thints like `JointActuator::AddInActuation(const Context<T>&, MultibodyForces<T>*)`

---

*[multibody/multibody_tree/multibody_plant/BUILD.bazel, line 25 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6Sc4AVa6JJXdWHBKG-:-L6SmdpZ1xmE_40kCecq:b-vc5soe) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/multibody_plant/BUILD.bazel#L25)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Given this is ostensibly about adding actuators, why are changes like these included?
</blockquote></details>

Good catch. I forgot to remove them before PR'ing. Done.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 144 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6Sawb09z00job1SuHu:-L6SoDRt8Ig66KNSlmiZ:b-dvuq3v) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L144)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Do you like the error message on this if no actuation was added?
</blockquote></details>

BTW of course I don't! good catch. Added appropriate demands and docs.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 383 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6S_1_LNzsfH1HORRQG:-L6SozhYxdzAoCwUC7YD:b-idelkc) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/multibody_plant/multibody_plant.h#L383)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW It isn't clear to me that `JointActuatorIndex` is a part of the public API. Where would the user get such a value to use as a meaningful index?

Also, how does this work with a joint with multiple dof? In the implementation of the forces, you note that only 1-dof forces are supported. So, that should probably be included elsewhere in the documentation (here, the constructor for `JointActuator`, etc. And it should likewise be enforced.)
</blockquote></details>

Updated doc here and in `AddJointAcutator()`. For now MBT support multiple dofs (nothing in the MBT code prevents it), I am just limiting to one dof in the MBP code for now to avoid additional logic we really don't need for now.

---

*[multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc, line 124 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6ScVdrjT1PK-_nuCYA:-L6Sqi0eHpiSo3T4PEjo:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc#L124)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

This should probably be preceded by an `ASSERT_GT(plant_->num_actutors(), 0);`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> A typo and, I hope, the last iteration on the documentation of `JointActuator`. This time I feel I have a stronger handle on reality and my comment is more to the point.

---

Reviewed 7 of 7 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[multibody/multibody_tree/joint_actuator.h, line 24 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6SWgdA0ViGebOk_ovO:-L6Ti_yL98-_ght_kVLr:b-73wl3u) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/joint_actuator.h#L24)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW I think the docs I added now in MBP help a bit. Otherwise, it'll become clearer when I add the computation of the B matrix.
</blockquote></details>

I'm rethinking my original comment (having actually read your comment to @edrumwri ). My confusion was thinking there was no `index()` method. Then I saw it inherits it from `MultibodyTreeElement` (which makes sense given the template parameters.)

So, now the concern is, who cares?  If a user has a reference to a `JointActuator` they have access to the `JointActuatorIndex` via the `index()` method. But why do they care? What can they use that index for? This discussion suggests its to map the index to the joint. But that they can really do by calling the `joint()` method.

The documentation down in MBP is much more illuminating. As I previously suggsted, the mapping from index to joint need not be discussed because it's handled implicitly by the actuator. However, there is *another* mapping that has immense value and *should* be discussed: the mapping from `JointActuatorIndex` to the field in the input actuator vector for this actuator's value. *That* is the mapping that I would emphasize here and not the mapping from index to joint.

---

*[multibody/multibody_tree/joint_actuator.h, line 35 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6SYga_2tL2VTILd_p8:-L6TiUprNArW3AxfCfEU:b-ruw8t8) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/joint_actuator.h#L35)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I thought about it, yes. I settled here just to provide a homogeneous API, all MBT modeling elements are templated on `<T>`. Also, this would allow us to add `Context` methods in here if we decide. Thints like `JointActuator::AddInActuation(const Context<T>&, MultibodyForces<T>*)`
</blockquote></details>

Sounds good.  (ANd I realized that my second list item failed to render as `actuator.GetJoint<T>();`)

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 383 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6S_1_LNzsfH1HORRQG:-L6Tk1E_Rcotg323T28H:b-95s9v7) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/multibody_plant/multibody_plant.h#L383)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Updated doc here and in `AddJointAcutator()`. For now MBT support multiple dofs (nothing in the MBT code prevents it), I am just limiting to one dof in the MBP code for now to avoid additional logic we really don't need for now.
</blockquote></details>

Reading this, I now have a greater appreciation for the `JointActuatorIndex`. I've modified my earlier comment to reflect the new understanding.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 387 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6Tk9rP0PcHJy74K0Px:-L6Tk9rP0PcHJy74K0Py:b-matjd) ([raw file](https://github.com/robotlocomotion/drake/blob/3c36a00c7f787212c8f17b39c62abca55d0c464c/multibody/multibody_tree/multibody_plant/multibody_plant.h#L387)):*
> ```objc
>   /// Returns a constant reference to the input port for external actuation.
>   /// This input port is a vector valued port, indexed by JointActuatorIndex.
>   /// A actuator's index can be obtained with JointActuator::index().
> ```

BTW s/A/An

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-:-L6TlCvlxJAakvZuqD3_:b-yqeaiy)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Reviewed 10 of 10 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[multibody/multibody_tree/joint_actuator.h, line 24 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6SWgdA0ViGebOk_ovO:-L6X_vCfJOBs37pjLvn1:b3vtjaj) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/joint_actuator.h#L24)):*
> However, there is another mapping that has immense value and should be discussed: the mapping from JointActuatorIndex to the field in the input actuator vector for this actuator's value. That is the mapping that I would emphasize here and not the mapping from index to joint.

True, but that documentation belongs in MBP, not MBT code. I reshuffled things a bit and added the method `JointActuator::AddInOneForce()` so that we can make things a bit more specific without unnecessary mention to indexes here. Essentially just more sugar.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
The question of `JointActuatorIndex` isn't quite closed (plus a couple of typos).

---

Reviewed 5 of 5 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[multibody/multibody_tree/joint_actuator.h, line 24 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6SWgdA0ViGebOk_ovO:-L6XeISX8EeHyVoZEV6-:b-d8d8dw) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/joint_actuator.h#L24)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

> However, there is another mapping that has immense value and should be discussed: the mapping from JointActuatorIndex to the field in the input actuator vector for this actuator's value. That is the mapping that I would emphasize here and not the mapping from index to joint.

True, but that documentation belongs in MBP, not MBT code. I reshuffled things a bit and added the method `JointActuator::AddInOneForce()` so that we can make things a bit more specific without unnecessary mention to indexes here. Essentially just more sugar.
</blockquote></details>

I love the fact that you buried the implementation mechanism such that I can have a `JointActuator` and generate forces without thinking about indices. But that seems to strictly benefit MBP's internal workings.

HOWEVER, this isn't complete. Ultimately, MBP will export an input port: the actuation signal. The `JointActuator` hidden index tells me exactly how to interpret the values in the input Vector. Surely *that* is still part of the API. Someone *outside* of of MBP will need to do that. What is your plan for that?

It *seemed* to me, that `JointActuator` was supposed to be that interface. I.e., I declare a bunch of joint actuators on MBP. I keep references to those `JointActuator`s that I've declared. I use them to set my output port that I plug into the MBP input port.  Do I have this wrong?

---

*[multibody/multibody_tree/joint_actuator.h, line 65 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6XefHbCt0GYbu7a1KU:-L6XefHbCt0GYbu7a1KV:b-bk33gm) ([raw file](https://github.com/robotlocomotion/drake/blob/d79eee6320f86f0c9899c53512df458515d4657e/multibody/multibody_tree/joint_actuator.h#L65)):*
> ```objc
>   /// @param[in] context
>   ///   The context storing the state and parameters for the model to which
>   ///   `this` joint belongs.
> ```

BTW Technically `this` isn't a joint.

---

*[multibody/multibody_tree/joint_actuator.h, line 72 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6XepD57wE7fIPCXq4s:-L6XepD57wE7fIPCXq4t:b-flk6xa) ([raw file](https://github.com/robotlocomotion/drake/blob/d79eee6320f86f0c9899c53512df458515d4657e/multibody/multibody_tree/joint_actuator.h#L72)):*
> ```objc
>   /// @param[in] joint_tau
>   ///   Generalized force corresponding to the degree of freedom indicated by
>   ///   `joint_dof` to be added into `forces`. Refere to the specific Joint
> ```

BTW s/Refere/Refer

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-:-L6Xi3jU3B9xPiK162ht:bvr3ozj)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 7 of 16 files reviewed at latest revision, 1 unresolved discussion.

---

*[multibody/multibody_tree/joint_actuator.h, line 24 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6SWgdA0ViGebOk_ovO:-L6cWOrKFo0BDnPTNEM9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/joint_actuator.h#L24)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I love the fact that you buried the implementation mechanism such that I can have a `JointActuator` and generate forces without thinking about indices. But that seems to strictly benefit MBP's internal workings.

HOWEVER, this isn't complete. Ultimately, MBP will export an input port: the actuation signal. The `JointActuator` hidden index tells me exactly how to interpret the values in the input Vector. Surely *that* is still part of the API. Someone *outside* of of MBP will need to do that. What is your plan for that?

It *seemed* to me, that `JointActuator` was supposed to be that interface. I.e., I declare a bunch of joint actuators on MBP. I keep references to those `JointActuator`s that I've declared. I use them to set my output port that I plug into the MBP input port.  Do I have this wrong?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Back to you @seancurtis-tri 

---

Review status: 7 of 16 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-:-L6cWm77a5W4g2axQgow:b5fkq3b)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
One style-guide issue and one last suggestion regarding notation.

---

Reviewed 9 of 9 files at r8.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[multibody/multibody_tree/joint_actuator.h, line 24 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6SWgdA0ViGebOk_ovO:-L6dHGLdoEWifKw9BGE7:b60m2z5) ([raw file](https://github.com/robotlocomotion/drake/blob/ccbcc4f3863c4f2fdb51a787e8b0b43d72d86f6c/multibody/multibody_tree/joint_actuator.h#L24)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

FYI I love what you've done. The only thing I lack is the fact that the class documentation doesn't allude to it. But I'm not holding it up for this.

---

*[multibody/multibody_tree/joint_actuator.h, line 88 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6d6eo7GUiutJ6kLhbv:-L6d6eo7GUiutJ6kLhbw:brar5ih) ([raw file](https://github.com/robotlocomotion/drake/blob/4e97eae5b3b2d09d2d0c30281a992f6c5ba1f530/multibody/multibody_tree/joint_actuator.h#L88)):*
> ```objc
> 
>   /// Given the actuation values u for `this` actuator, this method sets the
>   /// actuation vector `mbt_u` for the entire %MultibodyTree model to which this
> ```

BTW Leftover `mbt_u` doesn't match `u` below.

And I also need to clarify my original suggestion. My idea was that `u` is the collection of *all* inputs (i.e., `y = f(t, x, u; p)`. It spans more than just actuation inputs. So, that would make `u_a` the actuation inputs and `u_ai` the actuation vector for the ith actuator.

Otherwise, this is good.

---

*[multibody/multibody_tree/multibody_tree_topology.h, line 266 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6dEQ3WOW-PyutQZx_P:-L6dEQ3XSV-FzPr0JKXe:b-oiplr1) ([raw file](https://github.com/robotlocomotion/drake/blob/4e97eae5b3b2d09d2d0c30281a992f6c5ba1f530/multibody/multibody_tree/multibody_tree_topology.h#L266)):*
> ```objc
>       int start_index, int ndofs) :
>       index(joint_actuator_index),
>       actuator_index_start(start_index), num_dofs(ndofs) {}
> ```

BTW GSG requires each of these to on a separate line (because they don't all fit on a single line).

---

*[multibody/multibody_tree/multibody_tree_topology.h, line 280 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6dEiXQlBYwRJ3bVB_A:-L6dEiXQlBYwRJ3bVB_B:b-d8kc21) ([raw file](https://github.com/robotlocomotion/drake/blob/4e97eae5b3b2d09d2d0c30281a992f6c5ba1f530/multibody/multibody_tree/multibody_tree_topology.h#L280)):*
> ```objc
>   JointActuatorIndex index{0};
>   /// For an actuator in a MultibodyTree model, this index corresponds to the
>   /// first entry in the global array u containing all actuation values for the
> ```

FYI See my previous not (in `joint_actuation.h` about `u` vs `u_a` an `u_ai`.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 155 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-L6dFLM1dHLjVHfkBy9n:-L6dFLM1dHLjVHfkBy9o:b3x3ira) ([raw file](https://github.com/robotlocomotion/drake/blob/4e97eae5b3b2d09d2d0c30281a992f6c5ba1f530/multibody/multibody_tree/multibody_plant/multibody_plant.h#L155)):*
> ```objc
> 
>   /// Returns the total number of actuated degrees of freedom.
>   /// That is, the vector of actuation values u has this size.
> ```

FYI Same note: `u`, `u_a`, and `u_ai`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8147#-:-L6dHS0xR82xku4gOKDi:b4isoms)*
<!-- Sent from Reviewable.io -->
